Fixing VRF fluid control model heating issue in compressor speed decrease with temperature, TU heating rate much larger than system

yujiex opened the pull request:
Pull request overview
---------------------
 - Fixes #10096
 - Fixes #10159
 - This PR fixes the VRF Fluid control model issues during cold weather when system capacity cannot meet the heating demand.

## Background

We used the component-based VRF system to model the heat pump in a single-family home in New York.
Our observations from the EnergyPlus time series results:
- Issue 1: When it is very cold outdoors, the VRF outdoor unit compressor does not stay at the maximal speed to provide heating; instead the speed drops! 
- Issue 2: Despite the VRF capacity (Rated Evaporative Capacity {W}) being too small to meet the heating/cooling demand, the terminal unit still supplies enough heating/cooling, with capacity not limited by the condenser. The heating part will be fixed here. The cooling part will be fixed in a separate PR with a separate issue number.

## Issue 1: Compressor speed and system heating rate dropping at low temperature

### Issue 1 diagnosis

We observe that VRF heating rate keeps and compressor speed keeps decreasing as temperature decreases. The compressor's maximum possible speed is 6900. Example files we used are "US+SF+CZ4A+hp+crawlspace+IECC_2006_VRF.idf" and "USA_NY_New.York-John.F.Kennedy.Intl.AP.744860_TMY3.epw"

<img width="559" alt="image" src="https://github.com/NREL/EnergyPlus/assets/13421161/4d136bf5-779d-419a-952b-0d256d4b3269">

The compressor speed output variable corresponds to `thisVrf.CompActSpeed` in the code, which gets value from a temporary variable `CompSpdActual`, which is computed in function `VRFOU_CalcCompH(.)`. This function loops through speed levels from small to large and finds the appropriate speed level. If in the end, it cannot find the proper speed, that means the required load to be met is larger than the max speed level can supply. The compressor speed will be set to the maximum speed. However, in the current code, the load is upper-bounded before entering the function when the load is larger than the system capacity.

```
          // *VRF OU Compressor Simulation at heating mode: Specify the compressor speed and power consumption
          this->VRFOU_CalcCompH(state,
                                TU_HeatingLoad,   <---- The load needs to be met.
                                this->EvaporatingTemp,
                                Tdischarge,
                                h_IU_cond_out_ave,
                                this->IUCondensingTemp,
                                CapMinTe,
                                Tfs,
                                Pipe_Q_h,
                                Q_c_OU,
                                CompSpdActual,  <---- Compressor speed is computed here.
                                Ncomp_new);
```

The TU_HeatingLoad variable is plotted against outdoor temperature on 1/1. It reduces with outdoor temperature, which is surprising. 

<img width="518" alt="image" src="https://github.com/NREL/EnergyPlus/assets/13421161/1d2be2dc-d0f9-4a83-bbb4-15a8187276ad">

TU_HeatingLoad was computed as the sum of all zone heating coil load met by TU (the corresponding heating coil's TotalHeatingEnergyRate (thisDXCoil.TotalHeatingEnerghRate, computed as AirMassFlow * (OutletAirEnthalpy - InletAirEnthalpy)).

```
    for (NumTU = 1; NumTU <= NumTUInList; ++NumTU) {
        TU_CoolingLoad += state.dataHVACVarRefFlow->TerminalUnitList(TUListNum).TotalCoolLoad(NumTU);
        TU_HeatingLoad += state.dataHVACVarRefFlow->TerminalUnitList(TUListNum).TotalHeatLoad(NumTU);
        TUParasiticPower +=
            state.dataHVACVarRefFlow->VRFTU(state.dataHVACVarRefFlow->TerminalUnitList(TUListNum).ZoneTUPtr(NumTU)).ParasiticCoolElecPower +
            state.dataHVACVarRefFlow->VRFTU(state.dataHVACVarRefFlow->TerminalUnitList(TUListNum).ZoneTUPtr(NumTU)).ParasiticHeatElecPower;
        TUFanPower += state.dataHVACVarRefFlow->VRFTU(state.dataHVACVarRefFlow->TerminalUnitList(TUListNum).ZoneTUPtr(NumTU)).FanPower;
    }
```

It gets trimmed by CompEvaporatingCAPSpdMax, which is the evaporating capacity at the highest compressor speed

```
        // Calculate total refrigerant flow rate
        if (Q_h_TU_PL > CompEvaporatingCAPSpdMax + CompEvaporatingPWRSpdMax) {
            // Required load is beyond the max system capacity

            Q_h_TU_PL = CompEvaporatingCAPSpdMax;
            TU_HeatingLoad = CompEvaporatingCAPSpdMax;
            this->TUHeatingLoad = TU_HeatingLoad;
            ...
         }
```

The CompEvaporatingCAPSpdMax is computed like this. `OUCoolingCAPFT(NumOfCompSpdInput)` is the user input curve for the max speed level.

```
        CompEvaporatingCAPSpdMax = this->CoffEvapCap * this->RatedEvapCapacity *
                                   CurveValue(state, this->OUCoolingCAPFT(NumOfCompSpdInput), this->IUCondensingTemp, CapMaxTe);
```

Inside the function `VRFOU_CalcCompH` that computes the compressor speed, the capacity supplied by each speed level is re-calculated. The following figure compares the `CompEvaporatingCAPSpdMax` used to upper bound `TU_HeatingLoad` and the capacity supplied by the maximum compressor speed inside `VRFOU_CalcCompH`. This means, in the current EnergyPlus code, the load to be met is cut to the red dots, and then in the compressor speed calculation, the program would think a much lower compressor speed would meet the load since the blue dots are so much higher than the red dots. The difference between the red and blue dots is larger in the morning and at night when the outdoor temperature is lower. 

<img width="1029" alt="image" src="https://github.com/NREL/EnergyPlus/assets/13421161/80ffd43c-122f-4a3d-b0c1-33018a9b36a7">

### Issue 1 fix

- Remove the upper bounding of `TU_HeatingLoad`

- Create a separate `TU_HeatingLoad_actual = min(TU_HeatingLoad, CompEvaporatingCAPSpdMax + CompEvaporatingPWRSpdMax)`. This will be used in the piping correction calculation.
After further diagnosis, this `TU_HeatingLoad_actual` is not actually necessary as the terminal unit output is already limited prior to adding up to `TU_HeatingLoad`. This will be removed in the next release.

## Issue 2: terminal heating coil heating rate > VRF heating rate

For a single-zone building, terminal heating coil heating rate > VRF heating rate when the condenser side can't meet the heating load. Furthermore, the terminal unit heating rate does not change with different values of "Rated Evaporative Capacity {W}" in AirConditioner:VariableRefrigerantFlow:FluidTemperatureControl

<img width="1126" alt="image" src="https://github.com/NREL/EnergyPlus/assets/13421161/6dfa5ebb-0b7c-4d8b-974d-d493854c5d77">

### issue 2 diagnose

For both the original and the fluid-control VRF model, terminal unit capacity should have been constrained by condenser capacity, where a maximum allowable terminal unit capacity is calculated in `LimitTUCapacity(.)` for each VRF system. The maximum allowable terminal unit capacity is stored in `state.dataHVACVarRefFlow->MaxHeatingCapacity(VRFCond)` and passed into `SimDXCoil(.)` which is then passed into `CalcVRFHeatingCoil_FluidT(.)` as an optional argument `MaxHeatCap`. This quantity is used to constrain `TotCap` (no defrost) and `TotCapAdj` (with defrost). `QCoilReq = TotCap * PartLoadRatio` is fed into `ControlVRFIUCoil(.)`. This function computes the `OutletAirEnthalpy` and `FanSpdRatio` (used to calculate `AirMassFlow`). Terminal unit coil heating rate is computed as `AirMassFlow * (OutletAirEnthalpy - InletAirEnthalpy)`. Since `QCoilReq` is not subject to the constraint `MaxHeatCap` placed, the final Terminal unit coil heating rate is not connected to VRF condenser side capacity. This leads to the seemingly unlimited heating rate at terminal unit heating coils.

```
void CalcVRFHeatingCoil_FluidTCtrl(EnergyPlusData &state,
                                   CompressorOperation const CompressorOp,              // compressor operation; 1=on, 0=off
                                   int const DXCoilNum,                                 // the number of the DX heating coil to be simulated
                                   Real64 const PartLoadRatio,                          // sensible cooling load / full load sensible cooling capacity
                                   int const FanOpMode,                                 // Allows parent object to control fan mode
                                   ObjexxFCL::Optional<Real64 const> OnOffAirFlowRatio, // ratio of compressor on airflow to compressor off airflow
                                   ObjexxFCL::Optional<Real64 const> MaxHeatCap         // maximum allowed heating capacity
)
{
    if ((AirMassFlow > 0.0) && (CompressorOp == CompressorOperation::On) && (GetCurrentScheduleValue(state, thisDXCoil.SchedPtr) > 0.0) &&
        (PartLoadRatio > 0.0) && (OutdoorDryBulb > thisDXCoil.MinOATCompressor)) {

        TotCap = thisDXCoil.RatedTotCap(Mode); <---- TotCap is not constraint by MaxCap (terminal unit maximum allowed capacity)
        QCoilReq = PartLoadRatio * TotCap;  
        if (PartLoadRatio == 0.0) {
            AirMassFlowMin = state.dataHVACVarRefFlow->OACompOffMassFlow;
        } else {
            AirMassFlowMin = state.dataHVACVarRefFlow->OACompOnMassFlow;
        }

        // Call ControlVRFIUCoil to calculate: (1) FanSpdRatio, (2) coil inlet/outlet conditions, and (3) SH/SC
        ControlVRFIUCoil(state,
                         DXCoilNum,
                         QCoilReq,    <---- PartLoadRatio * TotCap, also not constraint
                         thisDXCoil.InletAirTemp,
                         thisDXCoil.InletAirHumRat,
                         thisDXCoil.CondensingTemp,
                         AirMassFlowMin,
                         FanSpdRatio,
                         OutletAirHumRat,
                         OutletAirTemp,
                         OutletAirEnthalpy,
                         ActualSH,
                         ActualSC);
        AirMassFlow = FanSpdRatio * thisDXCoil.RatedAirMassFlowRate(Mode);
        ...
       
        if (present(MaxHeatCap)) {
            TotCapAdj = min(MaxHeatCap, TotCap * HeatingCapacityMultiplier);
            TotCap = min(MaxHeatCap, TotCap);   <---- the constraint happened too late
        } else {
            TotCapAdj = TotCap * HeatingCapacityMultiplier;
        }
...
}
```

### Issue 2 fix: move the capacity adjustment upfront before `QCoilReq = TotCap * PartLoadRatio`

## Effect of the fix on an existing EnergyPlus example file

For the `VariableRefrigerantFlow_FluidTCtrl_5Zone.idf` file, the system capacity is reduced to 4000W to examine the condition where the system capacity cannot meet the heating load. 

### Whether it fixed issue 1

The compressor speed and the VRF heat pump heating rate are plotted against outdoor temperature for 1/1. The blue dots are before fixing, and the orange dots are after fixing. We can see after fixing, the compressor speed and VRF heating rate both stayed high during cold weather.

<img width="465" alt="image" src="https://github.com/NREL/EnergyPlus/assets/13421161/56904d48-08c2-4ed7-8f5d-11a2fb76b111">

<img width="421" alt="image" src="https://github.com/NREL/EnergyPlus/assets/13421161/5d0b4a68-e493-43d9-9197-3a73d59494ac">

### whether it fixed issue 2

Before fixing, terminal units have very high capacity (higher than VRF condenser side capacity) and are only running part of the time. The set point temperature is met for all 5 spaces.

<img width="732" alt="image" src="https://github.com/NREL/EnergyPlus/assets/13421161/1f875305-c271-4666-aa32-d850e0788406">

After fixing, terminal unit capacities are reduced and they're running all the time. The temperature of the five spaces is much lower than the set point. This is anticipated as the VRF should not have the capacity to meet the load.

<img width="732" alt="image" src="https://github.com/NREL/EnergyPlus/assets/13421161/49e706de-5634-48fb-af59-e5efa69ac07a">

### whether supplemental heating works properly

The following verification uses the example file "VariableRefrigerantFlow_FluidTCtrl_wSuppHeater_5Zone.idf", with the New York weather.

Before fixing issue 2, the supplemental heating coils have zero heating rate, as the terminal heating coils have unlimited capacity and are enough for the full heating load (which is wrong from previous reasoning). After fixing issue 2, supplemental heating coils have non-zero consumption now.

before fixing

<img alt="image" src="https://github.com/NREL/EnergyPlus/assets/13421161/54bdacb4-f767-41de-97fe-9998e05e0331">

After fixing

<img alt="image" src="https://github.com/NREL/EnergyPlus/assets/13421161/193871d2-ca1d-436b-a927-afab436721e0">

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

yujiex committed:
add test idf and epw, will be removed after issue is fixed

yujiex committed:
Fixing compressor spd calc with capped load, TU capacity not limited

yujiex committed:
Merge remote-tracking branch 'origin/develop' into VRFheatingFixHeatRateCompSpd

yujiex committed:
add test idfs to CMakeLists.txt

rraustad commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -17312,6 +17312,7 @@ void CalcVRFHeatingCoil_FluidTCtrl(EnergyPlusData &state,
                                                    state.dataDXCoils->DXCoil(thisDXCoil.CompanionUpstreamDXCoil).CoolingCoilRuntimeFraction));
         }
 
+        OutletAirEnthalpy = min(OutletAirEnthalpy, FullLoadOutAirEnth);
```
Why? Is this model not smart enough to calculate this correctly in the first place? There are 2 modes of operation in a coil. CyclingFan, or ConstantFan. For CyclingFan mode, the outlet condition is the full outlet condition and the air mass flow rate is the average. For ConstantFan mode, the outlet condition is the average of compressor off condition (the inlet) and compressor on condition (the outlet at full capacity). So what this says is I don't care what fan cycling mode is used, the model will just output the full load outlet enthalpy, which will always be lower than the average enthalpy based on PLR. So no, please do not do this.

yujiex replied:
I was trying to incorporate the output `state.dataHVACVarRefFlow->MaxHeatingCapacity(VRFCond)` from the LimitTUCapacity() into the calculation of OutletAirEnthalpy. The OutletAirEnthalpy was calculated in this function
```
void ControlVRFIUCoil(EnergyPlusData &state,
                      int const CoilIndex,     // index to VRFTU coil
                      Real64 const QCoil,      // coil load
                      Real64 const Tin,        // inlet air temperature
                      Real64 const Win,        // inlet air humidity ratio
                      Real64 const TeTc,       // evaporating or condensing temperature
                      Real64 const OAMassFlow, // mass flow rate of outdoor air
                      Real64 &FanSpdRatio,     // fan speed ratio: actual flow rate / rated flow rate
                      Real64 &Wout,            // outlet air humidity ratio
                      Real64 &Tout,            // outlet air temperature
                      Real64 &Hout,            // outlet air enthalpy
                      Real64 &SHact,           // actual SH
                      Real64 &SCact            // actual SC
)
```
In that function, it didn't involve the use of this maximum terminal unit capacity. The Hout (outlet air enthalpy) is computed like this

```
            FanSpdRatioMax = 1.0;
            auto f = [QCoilSenHeatingLoad, Ts_1, Tin, Garate, BF](Real64 FanSpdRto) {
                return FanSpdResidualHeat(FanSpdRto, QCoilSenHeatingLoad, Ts_1, Tin, Garate, BF);
            };
            General::SolveRoot(state, 1.0e-3, MaxIter, SolFla, Ratio1, f, FanSpdRatioMin, FanSpdRatioMax);
            // this will likely cause problems eventually, -1 and -2 mean different things
            if (SolFla < 0) Ratio1 = FanSpdRatioMax; // over capacity
            FanSpdRatio = Ratio1;
            CoilOnOffRatio = 1.0;

            Tout = Tin + (Ts_1 - Tin) * (1 - BF);
            Wout = Win;
            Hout = PsyHFnTdbW(Tout, Wout);
```

So I added in a bound for it, so that the terminal unit capacity won't be crazily large. Without limitation, the output will be like this for a single-zone building (see column F and H)
<img width="600" alt="image" src="https://github.com/NREL/EnergyPlus/assets/13421161/b20dea3f-c9ab-4b01-86a6-b8ef503e8c4e">

But maybe the way I limited it is not right? @rraustad

rraustad commented on src/EnergyPlus/HVACVariableRefrigerantFlow.cc:
```diff
@@ -11784,7 +11784,7 @@ void VRFCondenserEquipment::CalcVRFCondenser_FluidTCtrl(EnergyPlusData &state)
 
             Q_h_TU_PL = CompEvaporatingCAPSpdMax;
-            TU_HeatingLoad = CompEvaporatingCAPSpdMax;
+            // TU_HeatingLoad = CompEvaporatingCAPSpdMax;
```
I am not very familiar with this model, so I will ask that this be explained. Why is this commented out?

yujiex replied:
I'm still working on this part. It seems the capacity calculation here for CompEvaporatingCAPSpdMax and at the compressor speed calculation part in `VRFOU_CalcCompH(.)` are different. The following is a plot comparing the max-speed capacity at these two locations for January 1st. The red one is the CompEvaporatingCAPSpdMax here and the blue one is the max-speed capacity in `VRFOU_CalcCompH(.)`. Since VRFOU_CalcCompH(.) uses this TU_HeatingLoad as an input representing the load to be met (which is trimmed by a low capacity), the compressor speed is set to very low. Logically it seems weird to use a trimmed load to calculate compressor speed as the function already takes care of the case where load > capacity. So I commented it out for now but I'm not certain if this is the final solution. It might need some further adjustment as the Piping loss calculation might be affected.
<img width="468" alt="image" src="https://github.com/NREL/EnergyPlus/assets/13421161/b9caf5c3-83ab-496a-afca-27ee9f7d3b78">

The max-speed capacities are different at these two places (for the same timestep) because the two temperature inputs are different. They both use a curve like this 
`this->CoffEvapCap * this->RatedEvapCapacity * CurveValue(state, this->OUCoolingCAPFT(NumOfCompSpdInput), arg1, arg2)`. 

The two temperature args are like this

  | Compressor spd calc	| TU_heatingload = CompEvaporatingCAPSpdMax
-- | -- | --
Arg 1 name | T_discharge | This->IUCondensingTemp
Arg 1 value | saturation temperature of a refrigerant, a function of   pressure, GetSatTemperatureRefrig(.) | T_coil_in + (Tout – T_coil_in) / (1 – BFH) + DeltaT, BFH is a constant, T_coil_in and Tout are air temperatures at the indoor unit inlet and outlet
Arg 2 name | T_suction | CapMaxTe
Arg 2 value | T_coil_surf – deltaT. deltaT is a curve of outdoor unit superheating degrees (this->SH), the curve is from “Outdoor Unit Evaporating Temperature Function of Superheating Curve Name” | OutDryBulb – this->SH

The values of the two temperature arguments are like these in 1/1
![arg1](https://github.com/NREL/EnergyPlus/assets/13421161/3585af20-9129-4f0b-bd31-bff69ccdad2c)
![arg2](https://github.com/NREL/EnergyPlus/assets/13421161/120312bd-73ef-4bd7-9eb4-a1b976387de9)

rraustad replied:
I think some detailed plots of a zone TU when the OA temp drops dramatically in the same day will show whether this is working correctly or not. The other thing is to check on the PipingCorrection* calculation/reports to make sure those make sense.

yujiex committed:
Directly limit capacity by moving MaxHeatCap chunk up front

rraustad commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -17168,6 +17168,16 @@ void CalcVRFHeatingCoil_FluidTCtrl(EnergyPlusData &state,
+        } else {
+            TotCapAdj = TotCap * HeatingCapacityMultiplier;
+        }
+
```
There are only 2 calls to CalcVRFHeatingCoil_FluidTCtrl. Is there a reason why MaxHeatCap can't be passed in both calls?

yujiex replied:
The one called in InitDXCoil does not pass in a MaxHeatCap. I inserted a breakpoint there and saw that at that moment the output from `LimitTUCapacity`, `state.dataHVACVarRefFlow->MaxHeatingCapacity(1)`, was 0. It has not been properly updated yet. It would be bad to pass in 0 as the capacity limit. The `CalcVRFHeatingCoil_FluidTCtrl` inside `InitDXCoil` is triggered only once, so it won't ever have a non-zero `MaxHeatCap` input. Maybe that's why it's not passed in? A few other coil types also did the same thing.
<img width="1474" alt="image" src="https://github.com/NREL/EnergyPlus/assets/13421161/c02c9203-dc66-49f0-8245-5e4d3d11569e">

rraustad replied:
See the use of this variable:

    constexpr Real64 MaxCap(1.0e+20); // limit of zone terminal unit capacity

yujiex replied:
Or at the initialization step, would there already be a more sensible max capacity? rather than a 1.0e+20?

rraustad replied:
For initializing the TU coil after sizing, the TU does not need to be limited. The call with no max cap argument only happens 1 time. You could pass MaxCap there without any repercussions. And then that argument would no longer need to be optional. Just a thought.

rraustad replied:
What I mean is, instead of doing this:

        // MaxHeatCap passed from parent object VRF Condenser and is used to limit capacity of TU's to that available from condenser
        if (present(MaxHeatCap)) {
            TotCapAdj = min(MaxHeatCap, TotCap * HeatingCapacityMultiplier);
            TotCap = min(MaxHeatCap, TotCap);
        } else {
            TotCapAdj = TotCap * HeatingCapacityMultiplier;
        }

Just pass in MaxCap and you can reduce this code to:

            TotCapAdj = min(MaxHeatCap, TotCap * HeatingCapacityMultiplier);
            TotCap = min(MaxHeatCap, TotCap);

rraustad replied:
Or better yet, this:

        TotCap = min(MaxHeatCap, TotCap);

        // Calculate full load outlet conditions
        FullLoadOutAirEnth = InletAirEnthalpy + (TotCap * HeatingCapacityMultiplier) / AirMassFlow;

Although, I am not sure why TotCapAdj is only use here? Does total capacity include defrost or doesn't it?

yujiex replied:
Thanks for the clarification. Now I understand what you meant. Indeed if the MaxHeatCap argument is not optional and I pass in the `MaxCap` (a constant of value 1.0e+20), it won't need to be an optional argument. Is an optional argument preferred or do we try to avoid it? If we don't want the optional argument, should we pass in something like `numeric_limits<float>::infinity()` to be more explicit? Maybe @Myoldmopar and @amirroth could also chime in on this?

For the other question, it seems `TotCap` doesn't include defrost. It is computed like this
`TotCap = thisDXCoil.RatedTotCap(Mode) * TotCapFlowModFac * TotCapTempModFac;`

rraustad commented on src/EnergyPlus/DXCoils.cc:
```diff
@@ -17207,19 +17217,9 @@ void CalcVRFHeatingCoil_FluidTCtrl(EnergyPlusData &state,
-        } else {
-            TotCapAdj = TotCap * HeatingCapacityMultiplier;
-        }
-
```
It is probably wise to determine if TU capacity is limited before calculating coil performance.

yujiex replied:
indeed.

rraustad reviewed (commented):
I think you are on the right track here to correcting this.

yujiex commented:
> I think some detailed plots of a zone TU when the OA temp drops dramatically in the same day will show whether this is working correctly or not. The other thing is to check on the PipingCorrection* calculation/reports to make sure those make sense.

For the terminal unit output. I plotted a few against the outside temperature. I turned off the supplementary heating. When it is cold outdoors, the zone temperature drops below 22 (setpoint), the terminal unit heating rate is leveling off, runtime fraction is set to max. Roughly looks reasonable.

<img width="1591" alt="image" src="https://github.com/NREL/EnergyPlus/assets/13421161/8cf4dddb-c670-456f-abe7-520037cffb5c">

yujiex commented:
Hi @rraustad, I also have some questions regarding the VRF model (both types as they share some routines).

1. Whether the logic used in `LimitTUCapacity` function matches the industry's common practice

Currently, in EnergyPlus, the `LimitTUCapacity` function is used to prevent total terminal unit capacity from going beyond the condenser capacity. It reduces the capacity of the zones with the largest load. Is this a common practice in VRF system control?

2. Comparing autosize vs fixed low VRF capacity, some terminal unit has increased heating rate but decreased temperature.
 
We used the `testfiles/VariableRefrigerantFlow_5Zone.idf` file to examine how the capacity limiting works. The following shows the space temperature and the terminal unit heating rate when all terminal unit heating coil capacities and the VRF system capacity use autosize.

<img width="468" alt="image" src="https://github.com/NREL/EnergyPlus/assets/13421161/54dcaa4e-67af-4a8d-aaa9-96af802e8adf">

The following is when the "Gross Rated Heating Capacity {W}" in AirConditioner:VariableRefrigerantFlow is set to 4000W. In this case, the system cannot meet the load, so the terminal unit coil capacities got adjusted. Using the existing logic, all of them are set to equal.

<img width="468" alt="image" src="https://github.com/NREL/EnergyPlus/assets/13421161/fee8feb6-9386-42bb-82dd-94bccf1f10dc">

Comparing these two tables, we can see <ins>for space 2-1 and space 4-1, even if the heating rate is higher (900+) in the bottom table, the zone temperature is lower somehow.</ins>

rraustad commented:
The logic in LimitTUCapacity is just what I came up with at the time. I tried to figure out what a set of TUs would do if the outdoor unit capacity was limited. It may be that all TUs are proportionally limited, or something in between that and what it does now. The was no manufacturers input on this functionality, nor has there been any complaints about it.

The VRF capacity limiting scheme is supposed to do this:

![image](https://github.com/NREL/EnergyPlus/assets/5778156/ff709213-fd30-493c-86e4-6e895c050e96)

rraustad commented:
And you are right, Space2 and Space4 should have met the zone Tstat set point.

yujiex commented:
Thanks for the clarification @rraustad.

yujiex committed:
Remove upper bound TU_HeatingLoad, keep bounded in piping calc

nrel-bot-2b commented:
VRFheatingFixHeatRateCompSpd (yujiex) - x86_64-Linux-Ubuntu-22.04-gcc-11.4: OK (3367 of 3507 tests passed, 185 test warnings)

Messages:\n
 * 140 tests had: Table big diffs.
 * 182 tests had: Table small diffs.
 * 2 tests had: EIO diffs.
 * 2 tests had: ESO small diffs.
 * 2 tests had: MTR small diffs.
 * 4 tests had: ERR diffs.

Failures:\n

regression Test Summary
 * Passed: 658
 * Failed: 140


<a href='https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-f8676cf6183741c4f545dcc62d2e51cb9bcf78d8-x86_64-Linux-Ubuntu-22.04-gcc-11.4.html'>![Build Badge](http://img.shields.io/badge/build%20status-passing-green.png)</a> <a href='https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-f8676cf6183741c4f545dcc62d2e51cb9bcf78d8-x86_64-Linux-Ubuntu-22.04-gcc-11.4.html'>![Test Badge](http://img.shields.io/badge/tests%20passed-96.01%25-yellow.png)</a>

nrel-bot-2b commented:
VRFheatingFixHeatRateCompSpd (yujiex) - x86_64-Linux-Ubuntu-22.04-gcc-11.4-UnitTestsCoverage-Debug: OK (1908 of 1908 tests passed, 0 test warnings)






<a href='https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-f8676cf6183741c4f545dcc62d2e51cb9bcf78d8-x86_64-Linux-Ubuntu-22.04-gcc-11.4-UnitTestsCoverage-Debug.html'>![Build Badge](http://img.shields.io/badge/build%20status-passing-green.png)</a> <a href='https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-f8676cf6183741c4f545dcc62d2e51cb9bcf78d8-x86_64-Linux-Ubuntu-22.04-gcc-11.4-UnitTestsCoverage-Debug.html'>![Test Badge](http://img.shields.io/badge/tests%20passed-100.0%25-green.png)</a> <a href='https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-f8676cf6183741c4f545dcc62d2e51cb9bcf78d8-x86_64-Linux-Ubuntu-22.04-gcc-11.4-UnitTestsCoverage-Debug.html'>![Coverage Badge](http://img.shields.io/badge/coverage%20status-50.12%25-green.png)</a>

nrel-bot-2b commented:
VRFheatingFixHeatRateCompSpd (yujiex) - x86_64-Linux-Ubuntu-22.04-gcc-11.4-IntegrationCoverage-Debug: OK (782 of 782 tests passed, 0 test warnings)






<a href='https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-f8676cf6183741c4f545dcc62d2e51cb9bcf78d8-x86_64-Linux-Ubuntu-22.04-gcc-11.4-IntegrationCoverage-Debug.html'>![Build Badge](http://img.shields.io/badge/build%20status-passing-green.png)</a> <a href='https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-f8676cf6183741c4f545dcc62d2e51cb9bcf78d8-x86_64-Linux-Ubuntu-22.04-gcc-11.4-IntegrationCoverage-Debug.html'>![Test Badge](http://img.shields.io/badge/tests%20passed-100.0%25-green.png)</a> <a href='https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-f8676cf6183741c4f545dcc62d2e51cb9bcf78d8-x86_64-Linux-Ubuntu-22.04-gcc-11.4-IntegrationCoverage-Debug.html'>![Coverage Badge](http://img.shields.io/badge/coverage%20status-68.22%25-green.png)</a>

yujiex committed:
Merge remote-tracking branch 'origin/develop' into VRFheatingFixHeatRateCompSpd

Myoldmopar commented:
This is looking quite good with lots of review.  The CI results that are showing up here are not necessarily accurate because they occurred over the weekend when I was trying some CI updates.  I think pulling up with develop would get a fresh set of results and also a passing CppCheck run.  I won't do that in case there is anything left to do here.  Do you need to see more @rraustad?

rraustad commented:
The change here seems to fix the issue with reduced compressor speed when it should run near or at full speed. The issue with how LimitTUCapacity works is outside the scope but still needs to be fixed. And the clang format with a fresh pull from develop.

rraustad commented:
Oh, not clang, that was CppCheck as you mentioned.

yujiex committed:
Merge remote-tracking branch 'origin/develop' into VRFheatingFixHeatRateCompSpd

yujiex commented:
@rraustad and @Myoldmopar, I just merged in develop. There's still a regression diff on the VariableRefrigerantFlow_FluidTCtrl_wSuppHeater_5Zone.idf with an extra warning message in the feature branch. I'm investigating it. I will try to push after resolving that.

yujiex committed:
Correct default value for Stop running case

it should use the default value, default value of MaxHeatingCapacity and
MaxCoolingCapacity are MaxCap, not 0. Otherwise when 0 is input to the function,
it will make the TotCap 0.
PipingCorrectionCooling default value is 1.0, not 0.0

yujiex commented:
@Myoldmopar I've created a separate cooling issue about terminal cooling rate > VRF cooling rate: #10187. A separate PR will be created to fix that issue. This PR will only deal with the heating part. Initially, I thought cooling would use roughly the same fix but it seems more involved. Currently, I don't think there are regression diffs on the CI (apart from PERFLOGs which is not related), as CI only runs design day simulations, and those won't cover the time period where the system cannot meet the heating load. I added some evaluation of the impact on an existing EnergyPlus example file in the PR description section "Effect of the fix on an existing EnergyPlus example file".

Myoldmopar commented:
This was as clean as expected the last time CI ran.  I'll do a quick build/test here as well and see if this can go right in.

yujiex commented:
> This was as clean as expected the last time CI ran. I'll do a quick build/test here as well and see if this can go right in.

Thanks, @Myoldmopar

Myoldmopar commented:
Yep, all good, merging.  Thanks @yujiex
